File: LeetCode/Easy/953 - Verifying an Alien Dictionary/953 - Verifying an Alien Dictionary.js
/**
 * @param {string[]} words
 * @param {string} order
 * @return {boolean}
 */
 var isAlienSorted = function(words, order) {
    const fr = {};
    for (let i = 0; i < 26; ++i) fr[order[i]] = i;
    
    const n = words.length;
    for (let i = 1; i < n; ++i) {
        if (!compare(fr, words[i], words[i - 1])) {
            return false;
        }
    }
    
    return true;
};

var compare = function(fr, word1, word2) {
    const n1 = word1.length, n2 = word2.length;
    const n = Math.min(n1, n2);
    
    for (let i = 0; i < n; ++i) {
        if (fr[word1[i]] != fr[word2[i]]) {
            return (fr[word1[i]] > fr[word2[i]]);
        }
    }
    
    return (n1 >= n2);
};

File: LeetCode/Easy/953 - Verifying an Alien Dictionary/953 - Verifying an Alien Dictionary.cpp
class Solution {
public:
    bool isAlienSorted(vector<string>& words, string order) {
        vector<int> fr(26);
        for (int i = 0; i < 26; ++i) fr[order[i] - 'a'] = i;

        int n = words.size();
        for (int i = 1; i < n; ++i) {
            if (!compare(fr, words[i], words[i - 1])) {
                return false;
            }
        }
        return true;
    }
    
    bool compare(vector<int> fr, string word1, string word2) {
        int n1 = word1.size(), n2 = word2.size(), n = min(n1, n2);
        for (int i = 0; i < n; ++i) {
            if (fr[word1[i] - 'a'] != fr[word2[i] - 'a']) {
                return (fr[word1[i] - 'a'] > fr[word2[i] - 'a']);
            }
        }
        return (n1 >= n2);
    };
};